service-base/src/main/java/org/eclipse/hono/service/tenant/TenantService.java

```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.service.tenant;
+
+import java.net.HttpURLConnection;
+
+import org.eclipse.hono.util.TenantResult;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Handler;
+import io.vertx.core.Verticle;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * A service for keeping record of tenant information.
+ *
+ */
+public interface TenantService extends Verticle {
+
+    /**
+     * Gets tenant data by tenant ID.
+     *
+     * @param tenantId The ID of he tenant for which data is requested.
+     * @param resultHandler The handler to invoke with the result of the operation. If tenant exists, the
+     *            <em>status</em> will be {@link HttpURLConnection#HTTP_OK} and the <em>payload</em> will contain the
+     *            tenant object. Otherwise the status will be {@link HttpURLConnection#HTTP_NOT_FOUND}.
+     */
+    void get(String tenantId, Handler<AsyncResult<TenantResult>> resultHandler);
```

Comment:
please change all method signatures to return a Future instead of passing in a handler.
Also document that the returned Future only succeeds if the response contains the expected *success* status code and is failed with a `ServiceInvocationException` otherwise.

Reply:
@sophokles73 : I gave that a try, here is the outcome:


` get(tenantId).map(tenantData -> {

                    reply(tenantMsg, tenantData);
                    return (Void)null;
                }).recover(t -> {
                    reply(tenantMsg, (ServiceInvocationException) t);
                    return Future.failedFuture(t);
                });`

with some changes to the helper methods (which are straightforward).

Pros:
- we can eliminate the TenantResult class completely
- we have a more natural signature with returning Futures (instead of a handler as param). See #471 which covers the same for the HonoClient. It would be aligned with that strategy then as well.

Cons:
- the calling code is lengthier when calling the implementation methods (see above). But this is of no big importance to me (and it might be possible to improve this snippet further).

Summary: 
I think we should do that!
But I would not do it in this PR, but create an issue and then do that for all APIs (registration and credentials as well) then.
WDYT?

Reply:
I am sorry, I was erroneously assuming that we are talking about the `TenantClient` here. For the service interface and implementation, I think that the `TenantResult` is helpful because we can then handle the status code that is returned to the client in a uniform way, right?

Reply:
The TenantClient is not yet inside this PR and for the client IMHO the Future return value is definitely the way to go.

Regarding the service implementation of the Tenant API:
the prototypical implementation of the get operation resulted in exactly the same number of lines, but using the Future as the result object (instead of the TenantResult).
The status code is contained in the exception as error code (in case of failure), or STATUS_OK (in case of succeeded future).

So IMHO there is not much advantage left regarding the TenantResult object, and we would have one class less.
Is nothing of top priority in my eyes, though - much more important to the the Tenant API finished. Refactoring can be done later.
If we agree, I would just write an issue for that now.

Reply:
How does the `AmqpEndpoint` know which (success) status code needs to be transferred back to the client for each operation invoked?